src/send_future.rs

```diff
@@ -0,0 +1,180 @@
+use crate::manager::AddressMessage;
+use crate::receiver::Receiver;
+use crate::Disconnected;
+use flume::r#async::SendFut;
+use futures_core::future::BoxFuture;
+use futures_util::FutureExt;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A [`Future`] that represents the state of sending a message to an actor.
+///
+/// By default, a `SendFuture` will also wait until the handler has finished executing and resolve to the return value (see [`Handler::Return`](crate::Handler::Return)).
+/// This behaviour can be changed by calling [`recv_async`](SendFuture::recv_async).
+///
+/// When toggled to receive the return value asynchronously, this future will resolve once the message is successfully queued into the actor's mailbox. If the actors mailbox is bounded, this future will yield `Pending` until there is space in the mailbox. This allows an actor to exercise backpressure on its users.
+#[must_use = "Futures do nothing unless polled"]
+pub struct SendFuture<R, F, TRecvSyncMarker> {
+    inner: SendFutureInner<R, F>,
+    phantom: PhantomData<TRecvSyncMarker>,
+}
+
+/// Marker type for receiving the return value of a [`Handler`](crate::Handler) **synchronously**.
+pub enum ReceiveSync {}
+
+/// Marker type for receiving the return value of a [`Handler`](crate::Handler) **asynchronously**.
+pub enum ReceiveAsync {}
+
+enum SendFutureInner<R, F> {
+    Disconnected,
+    Sending(F),
+    Receiving(Receiver<R>),
+    Done,
+}
+
+impl<R, F> SendFuture<R, F, ReceiveSync> {
+    pub(crate) fn disconnected() -> Self {
+        Self {
+            inner: SendFutureInner::Disconnected,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Toggle this future to receive the return value asynchronously.
+    ///
+    /// Calling this function will change the [`Output`](Future::Output) of this [`Future`] from [`Handler::Return`](crate::Handler::Return) to [`Receiver<Handler::Return>`](Receiver<crate::Handler::Return>).
+    ///
+    /// A [`Receiver`] is a [`Future`] itself that can be used to await the execution of the [`Handler`](crate::Handler) separately from the queuing of the message.
+    pub fn recv_async(self) -> SendFuture<R, F, ReceiveAsync> {
```

Comment:
I like the concept of moving out the change in receiving behavior from `do_` to the `send` method itself with this kind of builder-style method. I wonder if there is a better name for it though - this method doesn't make the receiving async (like the name could imply), it rather just makes it happen separately from the send action. Ultimately, I'd just be worried the name would imply that `send` blocks to receive, which it does not

Reply:
> would imply that send blocks to receive, which it does not

Well, it does block the _task_ until the return value is ready. The mix of terminology here can be confusing though, I agree!

Reply:
> this method doesn't make the receiving async (like the name could imply), it rather just makes it happen separately from the send action.

When ignoring the specific semantics of `async/await` in Rust, the usage of "asynchronously" here is still correct I believe. It is like having:

1. An HTTP request which receives the response synchronously.
2. A "fire-and-forget" HTTP request that returns you a link from where you can fetch the result eventually.

I am happy to rename this to something else because the naming is clearly confusing due to the overloading of the term `asynchronously` with Rust's specific semantics.

A few other names I can think of are:

- `split_receiver`
- `into_receiver`
- `recv_separately`
- `queue_only` (because the future will resolve once the message is queued into the mailbox)

@zesterer Any opinion on the naming of this API?

Reply:
I had a go at renaming the function to `split_receiver` in https://github.com/Restioson/xtra/pull/74/commits/49e2f56de9a41dc2e45809e800898d97b0247467. Let me know what you think!

Reply:
That sounds nice to me I think - no more clash with Rust's sync vs async terminology :)